File: frontend/src/services/allocationService.ts
import api from './api';
import { Allocation, Employee, Demand, ProjectPhase, VacationPeriod, FixedAllocation } from '@/types/planner';

// ========== Types ==========

interface AllocationDto {
  id: string;
  employeeId: string;
  employeeName: string;
  demandId?: string | null;
  demandName?: string | null;
  projectId?: string | null;
  projectName?: string | null;
  month: number;
  year: number;
  hours: number;
  isLoan: boolean;
  sourceTeamId?: string;
  sourceTeamName?: string;
  allocationType?: string | null;
  createdAt: string;
}

interface AllocationEmployeeDto {
  id: string;
  name: string;
  role: string;
  teamId: string;
  teamName: string;
  dailyHours: number;
  vacations: VacationPeriodDto[];
  fixedAllocations: FixedAllocationSimpleDto[];
}

interface VacationPeriodDto {
  id: string;
  startDate: string;
  endDate: string;
}

interface FixedAllocationSimpleDto {
  id: string;
  name: string;
  hoursPerMonth: number;
}

interface AllocationDemandDto {
  id: string;
  projectId: string;
  projectName: string;
  projectColor: string;
  name: string;
  description?: string;
  teamId: string;
  startDate: string;
  endDate: string;
  totalHours: number;
  allocatedHours: number;
  status: string;
  phases: DemandPhaseSimpleDto[];
}

interface DemandPhaseSimpleDto {
  id: string;
  type: string;
  name?: string;
  startDate: string;
  endDate: string;
  isMilestone: boolean;
}

interface AllocationStatsDto {
  totalEmployees: number;
  totalDemands: number;
  totalHoursAllocated: number;
  loanAllocationsCount: number;
  overloadedEmployeesCount: number;
}

interface AllocationPageDataResponse {
  employees: AllocationEmployeeDto[];
  demands: AllocationDemandDto[];
  allocations: AllocationDto[];
  stats: AllocationStatsDto;
}

interface AllocationListResponse {
  allocations: AllocationDto[];
  totalCount: number;
}

interface CreateAllocationRequest {
  employeeId: string;
  demandId?: string | null;
  projectId?: string | null;
  month: number;
  year: number;
  hours: number;
  isLoan: boolean;
  sourceTeamId?: string;
  allocationType?: string | null; // null = demanda, "VACATION" = férias, "TRAINING" = treinamento
}

// ========== Exported Types ==========

export interface AllocationPageData {
  employees: Employee[];
  demands: Demand[];
  allocations: Allocation[];
  stats: {
    totalEmployees: number;
    totalDemands: number;
    totalHoursAllocated: number;
    loanAllocationsCount: number;
    overloadedEmployeesCount: number;
  };
}

// Constantes para tipos especiais de alocação (devem corresponder ao TimelineGrid)
const SPECIAL_ALLOCATION_TYPES = {
  VACATION: '__VACATION__',
  TRAINING: '__TRAINING__',
} as const;

// Mapeamento de IDs especiais para AllocationType do backend
const SPECIAL_TYPE_TO_ALLOCATION_TYPE: Record<string, string> = {
  [SPECIAL_ALLOCATION_TYPES.VACATION]: 'VACATION',
  [SPECIAL_ALLOCATION_TYPES.TRAINING]: 'TRAINING',
};

// Mapeamento inverso: AllocationType do backend para IDs especiais do frontend
const ALLOCATION_TYPE_TO_SPECIAL_ID: Record<string, string> = {
  'VACATION': SPECIAL_ALLOCATION_TYPES.VACATION,
  'TRAINING': SPECIAL_ALLOCATION_TYPES.TRAINING,
};

// ========== Helper Functions ==========

const parseDate = (dateString: string): Date => {
  const date = new Date(dateString);
  if (!dateString.includes('T')) {
    return new Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());
  }
  return date;
};

const mapDtoToAllocation = (dto: AllocationDto): Allocation => {
  // Se tiver allocationType, usar o ID especial correspondente
  const demandId = dto.allocationType 
    ? ALLOCATION_TYPE_TO_SPECIAL_ID[dto.allocationType] || dto.demandId || ''
    : dto.demandId || '';
  
  return {
    id: dto.id,
    employeeId: dto.employeeId,
    demandId: demandId,
    projectId: dto.projectId || '',
    month: dto.month,
    year: dto.year,
    hours: dto.hours,
    isLoan: dto.isLoan,
    sourceTeamId: dto.sourceTeamId,
    allocationType: dto.allocationType || undefined,
    // Informações adicionais para demandas de outros departamentos
    demandName: dto.demandName || undefined,
    projectName: dto.projectName || undefined,
  };
};

const mapDtoToEmployee = (dto: AllocationEmployeeDto): Employee => ({
  id: dto.id,
  name: dto.name,
  role: dto.role,
  teamId: dto.teamId,
  dailyHours: dto.dailyHours,
  vacations: dto.vacations.map(v => ({
    id: v.id,
    startDate: parseDate(v.startDate),
    endDate: parseDate(v.endDate),
  })),
  fixedAllocations: dto.fixedAllocations.map(f => ({
    id: f.id,
    name: f.name,
    hoursPerMonth: f.hoursPerMonth,
  })),
});

const mapDtoToDemand = (dto: AllocationDemandDto): Demand => ({
  id: dto.id,
  projectId: dto.projectId,
  name: dto.name,
  description: dto.description,
  teamId: dto.teamId,
  startDate: parseDate(dto.startDate),
  endDate: parseDate(dto.endDate),
  totalHours: dto.totalHours,
  allocatedHours: dto.allocatedHours,
  status: dto.status as Demand['status'],
  createdAt: new Date(),
  isNew: false,
  phases: dto.phases.map(p => ({
    id: p.id,
    type: p.type as ProjectPhase['type'],
    name: p.name,
    startDate: parseDate(p.startDate),
    endDate: parseDate(p.endDate),
    isMilestone: p.isMilestone,
  })),
  // Adicionar info do projeto para uso em componentes
  _projectName: dto.projectName,
  _projectColor: dto.projectColor,
});

// ========== API Service ==========

export const allocationService = {
  /**
   * Obtém todos os dados necessários para a página de alocação
   */
  async getPageData(teamId?: string, year?: number): Promise<AllocationPageData> {
    try {
      const params: Record<string, string | number> = {};
      if (teamId) params.teamId = teamId;
      if (year) params.year = year;

      const response = await api.get<AllocationPageDataResponse>('/api/allocations/page-data', { params });
      
      return {
        employees: response.data.employees.map(mapDtoToEmployee),
        demands: response.data.demands.map(mapDtoToDemand),
        allocations: response.data.allocations.map(mapDtoToAllocation),
        stats: {
          totalEmployees: response.data.stats.totalEmployees,
          totalDemands: response.data.stats.totalDemands,
          totalHoursAllocated: response.data.stats.totalHoursAllocated,
          loanAllocationsCount: response.data.stats.loanAllocationsCount,
          overloadedEmployeesCount: response.data.stats.overloadedEmployeesCount,
        },
      };
    } catch (error) {
      console.error('Erro ao buscar dados da página de alocação:', error);
      return {
        employees: [],
        demands: [],
        allocations: [],
        stats: {
          totalEmployees: 0,
          totalDemands: 0,
          totalHoursAllocated: 0,
          loanAllocationsCount: 0,
          overloadedEmployeesCount: 0,
        },
      };
    }
  },

  /**
   * Obtém todas as alocações com filtros
   */
  async getAllocations(teamId?: string, employeeId?: string, year?: number): Promise<Allocation[]> {
    try {
      const params: Record<string, string | number> = {};
      if (teamId) params.teamId = teamId;
      if (employeeId) params.employeeId = employeeId;
      if (year) params.year = year;

      const response = await api.get<AllocationListResponse>('/api/allocations', { params });
      return response.data.allocations.map(mapDtoToAllocation);
    } catch (error) {
      console.error('Erro ao buscar alocações:', error);
      return [];
    }
  },

  /**
   * Cria uma nova alocação
   */
  async createAllocation(allocation: Omit<Allocation, 'id'>): Promise<Allocation | null> {
    try {
      // Verifica se é um tipo especial de alocação (férias, treinamento, etc.)
      const isSpecialType = Object.values(SPECIAL_ALLOCATION_TYPES).includes(allocation.demandId as any);
      const allocationType = isSpecialType ? SPECIAL_TYPE_TO_ALLOCATION_TYPE[allocation.demandId] : null;

      const request: CreateAllocationRequest = {
        employeeId: allocation.employeeId,
        demandId: isSpecialType ? null : allocation.demandId,
        projectId: isSpecialType ? null : allocation.projectId,
        month: allocation.month,
        year: allocation.year,
        hours: allocation.hours,
        isLoan: allocation.isLoan ?? false,
        sourceTeamId: allocation.sourceTeamId,
        allocationType: allocationType,
      };

      const response = await api.post<AllocationDto>('/api/allocations', request);
      return mapDtoToAllocation(response.data);
    } catch (error) {
      console.error('Erro ao criar alocação:', error);
      return null;
    }
  },

  /**
   * Atualiza uma alocação existente
   */
  async updateAllocation(id: string, hours: number): Promise<Allocation | null> {
    try {
      const response = await api.put<AllocationDto>(`/api/allocations/${id}`, { hours });
      return mapDtoToAllocation(response.data);
    } catch (error) {
      console.error('Erro ao atualizar alocação:', error);
      return null;
    }
  },

  /**
   * Remove uma alocação
   */
  async deleteAllocation(id: string): Promise<boolean> {
    try {
      await api.delete(`/api/allocations/${id}`);
      return true;
    } catch (error) {
      console.error('Erro ao remover alocação:', error);
      return false;
    }
  },

  /**
   * Remove alocações por filtro
   */
  async deleteAllocationsByFilter(employeeId: string, demandId: string, month: number, year: number): Promise<boolean> {
    try {
      await api.delete('/api/allocations/by-filter', {
        params: { employeeId, demandId, month, year }
      });
      return true;
    } catch (error) {
      console.error('Erro ao remover alocações:', error);
      return false;
    }
  },
};

File: backend-api/src/ResourceFlow.Infrastructure/Services/AllocationService.cs
using Microsoft.EntityFrameworkCore;
using ResourceFlow.Application.DTOs.Allocations;
using ResourceFlow.Application.Interfaces;
using ResourceFlow.Domain.Entities;
using ResourceFlow.Infrastructure.Data;

namespace ResourceFlow.Infrastructure.Services;

public class AllocationService : IAllocationService
{
    private readonly ApplicationDbContext _context;

    public AllocationService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<AllocationPageDataResponse> GetAllocationPageDataAsync(Guid? teamId = null, int? year = null)
    {
        var currentYear = year ?? DateTime.Now.Year;

        // Buscar funcionários
        var employeesQuery = _context.Employees
            .Include(e => e.Department)
            .Include(e => e.Vacations)
            .Include(e => e.FixedAllocations)
            .Where(e => e.IsActive);

        if (teamId.HasValue)
        {
            employeesQuery = employeesQuery.Where(e => e.DepartmentId == teamId.Value);
        }

        var employees = await employeesQuery.OrderBy(e => e.Name).ToListAsync();

        // Buscar demandas (do ano atual ou que se estendem para o ano atual)
        var demandsQuery = _context.Demands
            .Include(d => d.Project)
            .Include(d => d.Team)
            .Include(d => d.Phases)
            .Where(d => d.Project!.IsActive);

        if (teamId.HasValue)
        {
            demandsQuery = demandsQuery.Where(d => d.TeamId == teamId.Value);
        }

        // Filtrar demandas do ano
        demandsQuery = demandsQuery.Where(d => 
            d.StartDate.Year <= currentYear && d.EndDate.Year >= currentYear);

        var demands = await demandsQuery.OrderBy(d => d.Name).ToListAsync();

        // Buscar alocações
        var allocationsQuery = _context.Allocations
            .Include(a => a.Employee)
            .Include(a => a.Demand)
            .Include(a => a.Project)
            .Include(a => a.SourceTeam)
            .Where(a => a.Year == currentYear);

        if (teamId.HasValue)
        {
            var employeeIds = employees.Select(e => e.Id).ToHashSet();
            var demandIds = demands.Select(d => d.Id).ToHashSet();
            allocationsQuery = allocationsQuery.Where(a => 
                employeeIds.Contains(a.EmployeeId) || demandIds.Contains(a.DemandId));
        }

        var allocations = await allocationsQuery.ToListAsync();

        // Calcular estatísticas
        var stats = new AllocationStatsDto
        {
            TotalEmployees = employees.Count,
            TotalDemands = demands.Count,
            TotalHoursAllocated = allocations.Sum(a => a.Hours),
            LoanAllocationsCount = allocations.Count(a => a.IsLoan),
            OverloadedEmployeesCount = 0 // Calculado no frontend baseado na capacidade
        };

        return new AllocationPageDataResponse
        {
            Employees = employees.Select(MapEmployeeToDto).ToList(),
            Demands = demands.Select(MapDemandToDto).ToList(),
            Allocations = allocations.Select(MapToDto).ToList(),
            Stats = stats
        };
    }

    public async Task<AllocationListResponse> GetAllocationsAsync(Guid? teamId = null, Guid? employeeId = null, int? year = null)
    {
        var query = _context.Allocations
            .Include(a => a.Employee)
            .Include(a => a.Demand)
            .Include(a => a.Project)
            .Include(a => a.SourceTeam)
            .AsQueryable();

        if (teamId.HasValue)
        {
            query = query.Where(a => 
                a.Employee.DepartmentId == teamId.Value || 
                a.Demand.TeamId == teamId.Value);
        }

        if (employeeId.HasValue)
        {
            query = query.Where(a => a.EmployeeId == employeeId.Value);
        }

        if (year.HasValue)
        {
            query = query.Where(a => a.Year == year.Value);
        }

        var allocations = await query.OrderBy(a => a.Year).ThenBy(a => a.Month).ToListAsync();

        return new AllocationListResponse
        {
            Allocations = allocations.Select(MapToDto).ToList(),
            TotalCount = allocations.Count
        };
    }

    public async Task<AllocationDto?> GetAllocationByIdAsync(Guid id)
    {
        var allocation = await _context.Allocations
            .Include(a => a.Employee)
            .Include(a => a.Demand)
            .Include(a => a.Project)
            .Include(a => a.SourceTeam)
            .FirstOrDefaultAsync(a => a.Id == id);

        return allocation == null ? null : MapToDto(allocation);
    }

    public async Task<AllocationDto> CreateAllocationAsync(CreateAllocationRequest request)
    {
        // Verificar se já existe uma alocação para este funcionário/demanda/mês/ano
        var existing = await _context.Allocations.FirstOrDefaultAsync(a =>
            a.EmployeeId == request.EmployeeId &&
            a.DemandId == request.DemandId &&
            a.Month == request.Month &&
            a.Year == request.Year);

        if (existing != null)
        {
            // Atualizar horas da alocação existente
            existing.Hours += request.Hours;
            existing.UpdatedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();
            
            return (await GetAllocationByIdAsync(existing.Id))!;
        }

        var allocation = new Allocation
        {
            Id = Guid.NewGuid(),
            EmployeeId = request.EmployeeId,
            DemandId = request.DemandId,
            ProjectId = request.ProjectId,
            Month = request.Month,
            Year = request.Year,
            Hours = request.Hours,
            IsLoan = request.IsLoan,
            SourceTeamId = request.SourceTeamId,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _context.Allocations.Add(allocation);
        
        // Atualizar horas alocadas na demanda
        var demand = await _context.Demands.FindAsync(request.DemandId);
        if (demand != null)
        {
            demand.AllocatedHours += request.Hours;
            demand.UpdatedAt = DateTime.UtcNow;
            
            // Atualizar status da demanda
            UpdateDemandStatus(demand);
        }

        await _context.SaveChangesAsync();

        return (await GetAllocationByIdAsync(allocation.Id))!;
    }

    public async Task<List<AllocationDto>> CreateBulkAllocationsAsync(BulkCreateAllocationRequest request)
    {
        var results = new List<AllocationDto>();
        
        foreach (var allocationRequest in request.Allocations)
        {
            var result = await CreateAllocationAsync(allocationRequest);
            results.Add(result);
        }

        return results;
    }

    public async Task<AllocationDto?> UpdateAllocationAsync(Guid id, UpdateAllocationRequest request)
    {
        var allocation = await _context.Allocations.FindAsync(id);
        if (allocation == null) return null;

        var hoursDiff = request.Hours - allocation.Hours;

        allocation.Hours = request.Hours;
        allocation.UpdatedAt = DateTime.UtcNow;

        // Atualizar horas alocadas na demanda
        var demand = await _context.Demands.FindAsync(allocation.DemandId);
        if (demand != null)
        {
            demand.AllocatedHours += hoursDiff;
            demand.UpdatedAt = DateTime.UtcNow;
            UpdateDemandStatus(demand);
        }

        await _context.SaveChangesAsync();

        return await GetAllocationByIdAsync(id);
    }

    public async Task<bool> DeleteAllocationAsync(Guid id)
    {
        var allocation = await _context.Allocations.FindAsync(id);
        if (allocation == null) return false;

        // Atualizar horas alocadas na demanda
        var demand = await _context.Demands.FindAsync(allocation.DemandId);
        if (demand != null)
        {
            demand.AllocatedHours -= allocation.Hours;
            if (demand.AllocatedHours < 0) demand.AllocatedHours = 0;
            demand.UpdatedAt = DateTime.UtcNow;
            UpdateDemandStatus(demand);
        }

        _context.Allocations.Remove(allocation);
        await _context.SaveChangesAsync();

        return true;
    }

    public async Task<bool> DeleteAllocationsByFilterAsync(Guid employeeId, Guid demandId, int month, int year)
    {
        var allocations = await _context.Allocations
            .Where(a => a.EmployeeId == employeeId && 
                       a.DemandId == demandId && 
                       a.Month == month && 
                       a.Year == year)
            .ToListAsync();

        if (!allocations.Any()) return false;

        var totalHours = allocations.Sum(a => a.Hours);

        // Atualizar horas alocadas na demanda
        var demand = await _context.Demands.FindAsync(demandId);
        if (demand != null)
        {
            demand.AllocatedHours -= totalHours;
            if (demand.AllocatedHours < 0) demand.AllocatedHours = 0;
            demand.UpdatedAt = DateTime.UtcNow;
            UpdateDemandStatus(demand);
        }

        _context.Allocations.RemoveRange(allocations);
        await _context.SaveChangesAsync();

        return true;
    }

    // ========== Helper Methods ==========

    private void UpdateDemandStatus(Demand demand)
    {
        if (demand.AllocatedHours == 0)
        {
            demand.Status = DemandStatus.Pending;
        }
        else if (demand.AllocatedHours >= demand.TotalHours)
        {
            demand.Status = DemandStatus.Allocated;
        }
        else
        {
            demand.Status = DemandStatus.Partial;
        }
    }

    private AllocationDto MapToDto(Allocation allocation)
    {
        return new AllocationDto
        {
            Id = allocation.Id,
            EmployeeId = allocation.EmployeeId,
            EmployeeName = allocation.Employee?.Name ?? string.Empty,
            DemandId = allocation.DemandId,
            DemandName = allocation.Demand?.Name ?? string.Empty,
            ProjectId = allocation.ProjectId,
            ProjectName = allocation.Project?.Name ?? string.Empty,
            Month = allocation.Month,
            Year = allocation.Year,
            Hours = allocation.Hours,
            IsLoan = allocation.IsLoan,
            SourceTeamId = allocation.SourceTeamId,
            SourceTeamName = allocation.SourceTeam?.Name,
            CreatedAt = allocation.CreatedAt
        };
    }

    private AllocationEmployeeDto MapEmployeeToDto(Employee employee)
    {
        return new AllocationEmployeeDto
        {
            Id = employee.Id,
            Name = employee.Name,
            Role = employee.Role,
            TeamId = employee.DepartmentId,
            TeamName = employee.Department?.Name ?? string.Empty,
            DailyHours = employee.DailyHours,
            Vacations = employee.Vacations.Select(v => new VacationPeriodDto
            {
                Id = v.Id,
                StartDate = v.StartDate,
                EndDate = v.EndDate
            }).ToList(),
            FixedAllocations = employee.FixedAllocations.Select(f => new FixedAllocationSimpleDto
            {
                Id = f.Id,
                Name = f.Name,
                HoursPerMonth = f.HoursPerMonth
            }).ToList()
        };
    }

    private AllocationDemandDto MapDemandToDto(Demand demand)
    {
        return new AllocationDemandDto
        {
            Id = demand.Id,
            ProjectId = demand.ProjectId,
            ProjectName = demand.Project?.Name ?? string.Empty,
            ProjectColor = demand.Project?.Color ?? string.Empty,
            Name = demand.Name,
            Description = demand.Description,
            TeamId = demand.TeamId,
            StartDate = demand.StartDate,
            EndDate = demand.EndDate,
            TotalHours = demand.TotalHours,
            AllocatedHours = demand.AllocatedHours,
            Status = MapStatusToString(demand.Status),
            Phases = demand.Phases.Select(p => new DemandPhaseSimpleDto
            {
                Id = p.Id,
                Type = MapPhaseTypeToString(p.Type),
                Name = p.Name,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                IsMilestone = p.IsMilestone
            }).ToList()
        };
    }

    private string MapStatusToString(DemandStatus status) => status switch
    {
        DemandStatus.Pending => "pending",
        DemandStatus.Partial => "partial",
        DemandStatus.Allocated => "allocated",
        DemandStatus.InProgress => "in-progress",
        DemandStatus.Completed => "completed",
        _ => "pending"
    };

    private string MapPhaseTypeToString(PhaseType type) => type switch
    {
        PhaseType.Construction => "construction",
        PhaseType.Homologation => "homologation",
        PhaseType.GoLive => "go-live",
        PhaseType.AssistedOperation => "assisted-operation",
        PhaseType.Maintenance => "maintenance",
        _ => "construction"
    };
}
